### PR TITLE
Test: Policy namespaces fixes

### PR DESCRIPTION
**Pull request by eloycoto**
- Move `Tests the same Policy in the different namespaces` into a new   
 context "Namespaces Policies"                                           
 - `Tests the same Policy in the different namespaces"                   
     - Moved to the namespace context.                                   
     - Use custom demo application instead of pods.                      
     - Move to `kubectl exec` instead of API pod creation that was       
     difficult to debug.                                                 

Related to #4292

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4304)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/Policies.go [hunk 1]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+			res.ExpectSuccess("unable to create namespace %q", namespace)
+
+			res = kubectl.Exec(fmt.Sprintf("kubectl label namespaces/%[1]s nslabel=%[1]s", namespace))
+			res.ExpectSuccess("cannot create namespace labels")
+
+			res = kubectl.Apply(demoManifest)
+			res.ExpectSuccess("unable to apply manifest %s")
+
+			res = kubectl.Apply(demoPath)
+			res.ExpectSuccess("unable to apply manifest %s")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: there is no corresponding string to `%s`

File: test/k8sT/Policies.go [hunk 2]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
-			})
+
+			res := kubectl.NamespaceCreate(namespace)
+			res.ExpectSuccess("unable to create namespace %q", namespace)
+
+			res = kubectl.Exec(fmt.Sprintf("kubectl label namespaces/%[1]s nslabel=%[1]s", namespace))
+			res.ExpectSuccess("cannot create namespace labels")
+
+			res = kubectl.Apply(demoManifest)
+			res.ExpectSuccess("unable to apply manifest %s")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: there is no corresponding string to `%s`

File: test/k8sT/Policies.go [hunk 3]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			res = kubectl.Apply(demoPath)
+			res.ExpectSuccess("unable to apply manifest %s")
+
+			ciliumPod, err = kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
+			Expect(err).Should(BeNil(), "cannot get CiliumPod")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: "cannot get CiliumPod on node %s", helpers.K8s1

File: test/k8sT/Policies.go [hunk 4]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			err := kubectl.WaitforPods(namespace, "-l zgroup=testapp", 300)
+			Expect(err).To(BeNil(), "testapp pods are not ready after timeout")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Since we wait for testapp pods to be ready in separate namespaces, include the namespace in the error message.

File: test/k8sT/Policies.go [hunk 5]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).To(BeNil(), "testapp pods are not ready after timeout")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Since we wait for testapp pods to be ready in separate namespaces, include the namespace in the error message.

File: test/k8sT/Policies.go [hunk 6]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
-						From: []networkingv1.NetworkPolicyPeer{{
-							NamespaceSelector: &metav1.LabelSelector{
-								MatchLabels: map[string]string{
-									"ns-name": namespace,
-								},
-							},
-						}},
-					}},
-				},
+		It("Tests the same Policy in the different namespaces", func() {
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: same policy in different namespaces

File: test/k8sT/Policies.go [hunk 7]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			By("Applying Policy in %q namespace", namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: include which policy you are applying in `By`

File: test/k8sT/Policies.go [hunk 8]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, policy, helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "L3 Policy cannot be applied in %q namespace", namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: This error message duplicates the one on line 814 - differentiate the two for easier debugging.

File: test/k8sT/Policies.go [hunk 9]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			By("Applying Policy in default namespace")
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, l3Policy, helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "L3 Policy cannot be applied in %q namespace", helpers.DefaultNamespace)
+
+			By("Testing %q namespace", namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Testing what in the namespace? Make this more specific.

File: test/k8sT/Policies.go [hunk 10]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+				namespace, appPodsNS[helpers.App2],
+				helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIPNS)))
+			res.ExpectSuccess("%q cannot curl service", appPods[helpers.App2])
+
+			res = kubectl.ExecPodCmd(
+				namespace, appPodsNS[helpers.App3],
+				helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIPNS)))
+			res.ExpectFail("%q can curl to service", appPods[helpers.App3])
+
+			By("Testing default namespace")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Testing what in the namespace? Make this more specific.

File: test/k8sT/Policies.go [hunk 11]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+				helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIP)))
+			res.ExpectSuccess("%q cannot curl clusterIP %q", appPods[helpers.App2], clusterIP)
+
+			res = kubectl.ExecPodCmd(
+				helpers.DefaultNamespace, appPods[helpers.App3],
+				helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIP)))
+			res.ExpectFail("%q can curl to %q", appPods[helpers.App3], clusterIP)
+		})
+
+		It("Network Policy namespace selector", func() {
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: This is not clear enough. What is this testing? Make the string reflect what this tests overall.

File: test/k8sT/Policies.go [hunk 12]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+			By("Applying Policy in %q namespace", namespace)
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, netpolNsSelector, helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "Policy cannot be applied")
+
+			for _, pod := range []string{helpers.App2, helpers.App3} {
+				res := kubectl.ExecPodCmd(
+					helpers.DefaultNamespace, appPods[pod],
+					helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIPNS)))
+				res.ExpectFail("%q can curl to service and it should not. NS policy should be applied", appPods[pod])
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: NS policy should be applied --> policy restricts communication across namespaces (or something more specific).

File: test/k8sT/Policies.go [hunk 13]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+				res = kubectl.ExecPodCmd(
+					namespace, appPodsNS[pod],
+					helpers.CurlFail(fmt.Sprintf("http://%s/public", clusterIPNS)))
+				res.ExpectSuccess("%q cannot curl service", appPodsNS[pod])
 			}
 
-			err := kubectl.ApplyNetworkPolicyUsingAPI(namespace, policyBeforeCreate)
-			Expect(err).To(BeNil(), "cannot apply network policy")
+			By("Delete policy in %q namespace", namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Nit - specify which policy is being deleted

File: test/k8sT/Policies.go [hunk 14]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
 	})
 
-	Context("KubernetesNetworkPolicy between server and client", func() {
+	Context("Namespaces policies", func() {
 
 		var (
-			policy = "allow-ns-b-via-namespace-selector"
+			err         error
+			ciliumPod   string
+			namespace   = "second"
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Rename this to "secondNamespace"

File: test/k8sT/Policies.go [hunk 15]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+			appPods     map[string]string
+			appPodsNS   map[string]string
+			clusterIP   string
+			clusterIPNS string
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Rename this variable to be something like "app1SecondNSClusterIP" - the variable name in its current form is not evocative of what it actually represents, and makes the test hard to follow.

File: test/k8sT/Policies.go [hunk 16]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+
+			demoPath         = helpers.ManifestGet("demo.yaml")
+			l3Policy         = helpers.ManifestGet("l3_l4_policy.yaml")
+			netpolNsSelector = fmt.Sprintf("%s -n %s", helpers.ManifestGet("netpol-namespace-selector.yaml"), namespace)
+			policy           = fmt.Sprintf("%s -n %s", l3Policy, namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Rename this variable. "policy" is quite vague. Please try to make the variable more expressive of what type of policy it represents.

File: test/k8sT/Policies.go [hunk 17]
```diff
@@ -827,78 +736,157 @@ EOF`, k, v)
...
+			err         error
+			ciliumPod   string
+			namespace   = "second"
+			appPods     map[string]string
+			appPodsNS   map[string]string
+			clusterIP   string
+			clusterIPNS string
+
+			demoPath         = helpers.ManifestGet("demo.yaml")
+			l3Policy         = helpers.ManifestGet("l3_l4_policy.yaml")
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: This variable is called l3Policy, yet the file is l3l4Policy; rename accordingly.

File: test/helpers/kubectl.go [hunk 18]
```diff
@@ -648,72 +647,6 @@ func (kub *Kubectl) WaitCleanAllTerminatingPods() error {
 	return err
 }
 
-// ApplyNetworkPolicyUsingAPI applies a Kubernetes network policy using the API
```
Comment (ianvernon) [inline test/helpers/kubectl.go:651]: <3

**Review by ianvernon — CHANGES_REQUESTED** [review]
Please address my comments. In the future, when adding tests to the CI, I would like to have the following to make it easier to be reviewed:]

* Describe the test topology in the commit. 
* Describe how policy is being enforced: what exactly is the test covering w.r.t. policy enforcement?
* Is there positive/negative testing being done w.r.t. the policy?

It would be even better if these comments were added inline to make the code more readable. I want to have this approach for policy-related tests.

EDIT: one additional comment I have is that the PR title is vague. What exactly is being fixed? It seems to be more of a refactoring to me. While this didn't impact my ability to review, it would be helpful in the future to have titles that are more evocative of what the PR is doing than the generic "fixes".

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/Policies.go [hunk 19]
```diff
@@ -827,78 +735,158 @@ EOF`, k, v)
...
+			res = kubectl.Apply(demoPath)
+			res.ExpectSuccess("unable to apply manifest")
+
+			err := kubectl.WaitforPods(namespace, "-l zgroup=testapp", 300)
+			Expect(err).To(BeNil(),
+				"testapp pods are not ready after timeout in namspace %q", namespace)
+
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).To(BeNil(),
+				"testapp pods are not ready after timeout in namespace", helpers.DefaultNamespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: There is no `%s` corresponding to `helpers.DefaultNamespace`

File: test/k8sT/Policies.go [hunk 20]
```diff
@@ -827,78 +735,158 @@ EOF`, k, v)
...
+			Expect(err).Should(BeNil(),
+				"%q Policy cannot be applied in %q namespace", l3l4PolicySecondNS, namespace)
+
+			By("Applying Policy in default namespace")
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, l3L4Policy, helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(),
+				"%q Policy cannot be applied in %q namespace", l3L4Policy, helpers.DefaultNamespace)
+
+			By("Testing connectivity on %q namespace", namespace)
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: Nit: on --> in

File: test/k8sT/Policies.go [hunk 21]
```diff
@@ -827,78 +735,158 @@ EOF`, k, v)
 		})
 	})
 
-	Context("KubernetesNetworkPolicy between server and client", func() {
+	Context("Namespaces policies", func() {
 
 		var (
-			policy = "allow-ns-b-via-namespace-selector"
+			err               error
+			namespace         = "second"
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: This variable still has not been renamed despite my prior request to do so. Please address my prior comment.

**Review by ianvernon — CHANGES_REQUESTED** [review]
@eloycoto a few comments from my last review, and my requests for more in-depth descriptions of the test topology have not been addressed. 
Please do the following:

* Describe the test topology in the commit.
* Describe how policy is being enforced: what exactly is the test covering w.r.t. policy enforcement?
* Is there positive/negative testing being done w.r.t. the policy? Provide evidence of this in the PR description.

It would be even better if these comments were added inline to make the code more readable. I want to have this approach for policy-related tests so that when people look at the tests they can easily understand what is going on. See `test/runtime/Policies.go:It(L3-Dependent L7 Egress)` for this. 

I realize that I have not asked for this before this PR, but I think this will make our tests much more approachable. Let me know what you think about this.

**Comment by eloycoto** [comment]
@ianvernon all should be ok. 

Regards

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/Policies.go [hunk 22]
```diff
@@ -827,78 +735,167 @@ EOF`, k, v)
...
+				helpers.KubeSystemNamespace, netpolNsSelector, helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "Policy cannot be applied")
+
+			for _, pod := range []string{helpers.App2, helpers.App3} {
+				// Make sure that the Default namespace can NOT connect to
+				// second namespace.
+				res := kubectl.ExecPodCmd(
+					helpers.DefaultNamespace, appPods[pod],
+					helpers.CurlFail(fmt.Sprintf("http://%s/public", secondNSclusterIP)))
+				res.ExpectFail("%q can curl to service, policiy is not blocking"+
```
Comment (ianvernon) [inline test/k8sT/Policies.go]: policiy --> policy

**Review by ianvernon — APPROVED** [review]
LGTM, apart from one small, non-blocking typo fix needed. Does not need a request for review from me again.

**Comment by eloycoto** [comment]
test-me-please
